ShellPkg/Application/Shell/Shell.c

```diff
@@ -1326,7 +1330,7 @@ DoStartupScript (
     }
 
     Status = RunShellCommand (FileStringPath, &CalleeStatus);
-    if (ShellInfoObject.ShellInitSettings.BitUnion.Bits.Exit == TRUE) {
+    if ((!EFI_ERROR (Status)) && (ShellInfoObject.ShellInitSettings.BitUnion.Bits.Exit == TRUE)) {

```

Comment:
`ShellCommandRegisterExit()` seems to maintain an internal state. It might be necessary to call the function even if `RunShellCommand()` failed. Not 100% sure, but would it be possible to check ?

Reply:
On this one I think we don't want to call `ShellCommandRegisterExit()` because if `RunShellCommand()` fails here, we didn't launch the startup script, so there is no currently running script to register an exit from. I think if you did register exit here, it could cause the entire shell to exit, since it is the currently running program. Or, more likely, it could cause the next script run to prematurely exit, since that global state has been set saying we want to exit. Let me know if you agree here or I am missing something.

Reply:
FWIU, we are talking here about the Shell application hosting the startup script to be launched.
I.e. it is possible to run from a first Shell instance (let's call it Shell0):
```
fs0:Shell.efi -exit fs0:Acpiview.efi
```

which will try to launch `Acpiview.efi` through a new Shell (let's call it Shell1). If for some reason this fails and the `-exit` parameter was given, I think we should still exit this Shell1 to go back to Shell0.

In any case, failing here doesn't seem to mean that the application (Acpiview.efi) or the startup script failed, as the exit status for this should be in `CalleeStatus`. Failing should be due here to parsing the command line, or memory allocation while parsing the command line

Reply:
Okay, sure I will update this to still call ShellCommandRegisterExit()